rs/crypto/ecdsa_secp256r1/tests/tests.rs

```diff
@@ -271,3 +272,94 @@ NRLvCGaIxJfchxpjcCysTG12MfKOf6/Phw==
         SAMPLE_SECP256R1_5915_PEM
     );
 }
+
+#[test]
+fn private_derivation_is_compatible_with_public_derivation() {
+    use rand::Rng;
+
+    let rng = &mut reproducible_rng();
+
+    fn random_path<R: Rng>(rng: &mut R) -> DerivationPath {
+        let l = 1 + rng.gen::<usize>() % 9;
+        let path = (0..l).map(|_| rng.gen::<u32>()).collect::<Vec<u32>>();
+        DerivationPath::new_bip32(&path)
+    }
+
+    for _ in 0..100 {
+        let master_sk = PrivateKey::generate_using_rng(rng);
+        let master_pk = master_sk.public_key();
+
+        let path = random_path(rng);
+
+        let chain_code = rng.gen::<[u8; 32]>();
+
+        let (derived_pk, cc_pk) = master_pk.derive_subkey_with_chain_code(&path, &chain_code);
+
+        let (derived_sk, cc_sk) = master_sk.derive_subkey_with_chain_code(&path, &chain_code);
+
+        assert_eq!(
+            hex::encode(derived_pk.serialize_sec1(true)),
+            hex::encode(derived_sk.public_key().serialize_sec1(true))
+        );
+
+        assert_eq!(hex::encode(cc_pk), hex::encode(cc_sk));
+
+        let msg = rng.gen::<[u8; 32]>();
+        let derived_sig = derived_sk.sign_message(&msg);
+
+        assert!(derived_pk.verify_signature(&msg, &derived_sig));
+    }
+}
+
```

Comment:
For some key derivations, we also have a test `private_derivation_also_works_for_derived_keys`: would that be applicable here as well?

Reply:
That test was written for and primarily relevant for Ed25519 because there (due to how Ed25519 secret keys are defined, as a seed that is hashed to produce the key), you can't perform an additive key derivation with a standard Ed25519 key. Instead you have to use the expanded key, which is the scalar + derandomization secret which are derived from the private seed by hashing it with SHA-512. So the return value of `derive_key` on the Ed25519 private key returns a completely different type, which cannot be serialized (since doing so would require computing a SHA-512 preimage) but can only sign. However it can (and does) support further derivation if required, and this test checks that derivation algorithms of `ed25519::PrivateKey` and `ed25519::DerivedPrivateKey` behave the same.

That said the same test does logically make sense here; for all derivation paths, chain codes, and keys, deriving the entire path at once is identical to deriving twice with the path split into two. So I imported the test.

Reply:
👍